connect/mirror/src/test/java/org/apache/kafka/connect/mirror/OffsetSyncStoreTest.java

```diff
@@ -57,52 +57,96 @@ public void testOffsetTranslation() {
+            assertEquals(OptionalLong.empty(), store.translateDownstream(null, tp, 200));
 
             // read a sync during startup
             store.sync(tp, 100, 200);
-            assertEquals(OptionalLong.empty(), store.translateDownstream(tp, 0));
-            assertEquals(OptionalLong.empty(), store.translateDownstream(tp, 100));
-            assertEquals(OptionalLong.empty(), store.translateDownstream(tp, 200));
+            assertEquals(OptionalLong.empty(), store.translateDownstream(null, tp, 0));
+            assertEquals(OptionalLong.empty(), store.translateDownstream(null, tp, 100));
+            assertEquals(OptionalLong.empty(), store.translateDownstream(null, tp, 200));
 
             // After the store is started all offsets are visible
             store.start();
-            assertEquals(OptionalLong.of(-1), store.translateDownstream(tp, 0));
-            assertEquals(OptionalLong.of(200), store.translateDownstream(tp, 100));
-            assertEquals(OptionalLong.of(201), store.translateDownstream(tp, 200));
+            assertEquals(OptionalLong.of(-1), store.translateDownstream(null, tp, 0));
+            assertEquals(OptionalLong.of(200), store.translateDownstream(null, tp, 100));
+            assertEquals(OptionalLong.of(201), store.translateDownstream(null, tp, 200));
         }
     }
 
     @Test
     public void testNoTranslationIfNoOffsetSync() {
         try (FakeOffsetSyncStore store = new FakeOffsetSyncStore()) {
             store.start();
-            assertEquals(OptionalLong.empty(), store.translateDownstream(tp, 0));
+            assertEquals(OptionalLong.empty(), store.translateDownstream(null, tp, 0));
         }
     }
+
+    @Test
+    public void testPastOffsetTranslation() {
+        try (FakeOffsetSyncStore store = new FakeOffsetSyncStore()) {
+            long maxOffsetLag = 10;
+            int offset = 0;
+            for (; offset <= 1000; offset += maxOffsetLag) {
+                store.sync(tp, offset, offset);
+            }
+            store.start();
+
+            // After starting but before seeing new offsets, only the latest startup offset can be translated
+            assertSparseSync(store, 1000, -1);
+
+            for (; offset <= 2000; offset += maxOffsetLag) {
+                store.sync(tp, offset, offset);
+            }
+
+            // After seeing new offsets, we still cannot translate earlier than the latest startup offset
+            assertSparseSync(store, 1000, -1);
+
+            // We can translate offsets between the latest startup offset and the latest offset with variable precision
+            // Older offsets are less precise and translation ends up farther apart
+            assertSparseSync(store, 1030, 1000);
+            assertSparseSync(store, 1540, 1030);
+            assertSparseSync(store, 1800, 1540);
+            assertSparseSync(store, 1920, 1800);
+            assertSparseSync(store, 1990, 1920);
+            assertSparseSync(store, 2000, 1990);
```

Comment:
Where do the argument values for these calls come from?

Naively, it looks like this test might have been written by observing the behavior of the `OffsetSyncStore` class (however arbitrary or even incorrect it may be) and then adding assertions that match the observed behavior, rather than identifying what correct values would look like and then ensuring that the `OffsetSyncStore` class's behavior gives those values.

Separately but possibly related, would it help to have a method that ensures that the invariants for the syncs stored by the `OffsetSyncStore` hold? I drafted something like this and it seems to work nicely:

In `OffsetSyncStoreTest`:
```java
private void assertSparseSyncInvariant(FakeOffsetSyncStore store, TopicPartition topicPartition) {
    for (int j = 0; j < Long.SIZE; j++) {
        for (int i = 0; i <= j; i++) {
            long jUpstream = store.syncFor(topicPartition, j).upstreamOffset();
            long iUpstream = store.syncFor(topicPartition, i).upstreamOffset();
            assertTrue(
                    jUpstream <= iUpstream,
                    "Upstream offset " + iUpstream + " at position " + i
                            + " should be at least as recent as offset " + jUpstream + " at position " + j
            );
            long iUpstreamBound = jUpstream + (1L << j);
            if (iUpstreamBound < 0)
                continue;
            assertTrue(
                    iUpstream < iUpstreamBound,
                    "Upstream offset " + iUpstream + " at position " + i
                    + " should be no greater than " + iUpstreamBound + " (" + jUpstream + " + 2^" + j + ")"
            );
        }
    }
}
```

(this method can be invoked after ever call to `OffsetSyncStore::sync`)

In `OffsetSyncStore`:
```java
// For testing
OffsetSync syncFor(TopicPartition topicPartition, int syncIdx) {
    OffsetSync[] syncs = offsetSyncs.get(topicPartition);
    if (syncs == null)
        throw new IllegalArgumentException("No syncs present for " + topicPartition);
    if (syncIdx >= syncs.length)
        throw new IllegalArgumentException(
                "Requested sync " + (syncIdx + 1) + " for " + topicPartition
                + " but there are only " + syncs.length + " syncs available for that topic partition"
        );
    return syncs[syncIdx];
}
```

Reply:
> Where do the argument values for these calls come from?

I agree that these numbers are quite opaque looking at them again. They are the next multiple of 10 after indexes which are very divisible by 2:
```
1030 -> 1024 -> 2^10
1540 -> 1536 -> 2^10 + 2^9
1780 -> 1792 -> 2^10 + 2^9 + 2^8
1920 -> 1920 -> 2^10 + 2^9 + 2^8 + 2^7
1990 -> 1984 -> 2^10 + 2^9 + 2^8 + 2^7 + 2^6
2000 -> 2000 -> 2^10 + 2^9 + 2^8 + 2^7 + 2^6 +     + 2^4
```
They are multiples of 10 because of the maxOffsetLag = 10, which does not know anything about this scheme and instead only emits syncs every 10 offsets.

Also note that the pattern skips 2^5 = 32 because 1984 + 32 = 2016 which is beyond the end of the test.

> Separately but possibly related, would it help to have a method that ensures that the invariants for the syncs stored by the OffsetSyncStore hold?

Thanks for the new assertions! i've added them to the test as you described.

Reply:
With the new constraints, i don't know how to describe how to derive the latest state of the sync store in a top-down fashion like I was with the bit masking algorithm. Effectively, I don't know how to derive these constants other than by running the algorithm and observing the results.

> rather than identifying what correct values would look like and then ensuring that the OffsetSyncStore class's behavior gives those values.

Looking at the evolving state of the array, I can tell that the current invariants do produce desirable behavior:
* The values are monotonic (always going further back in time as the index increases)
* More recent elements in the array are closer together than later elements
* The number of distinct values only monotonically increases (no cascading clears like the old algorithm)

The last point wasn't the case, even with the updated invariants. I had an off-by-one error which was causing cascading clearing, which I noticed with an explicit test, and fixed with just a tweak to the invariants.
So thanks for pointing out that this test was inadequate, that motivated me to add another test which found a bug!

If you have any other suggestions for how to effectively test this, I'm happy to include them.

Reply:
Thanks for the explanation! I like the values here. They illustrate the granularity with which we can translate offsets (and how that granularity decreases as we have to use older and older data) quite nicely.